src/databricks/labs/ucx/source_code/graph.py

```diff
@@ -134,19 +134,31 @@ def root_dependencies(self) -> set[Dependency]:
 
         def add_to_dependencies(graph: DependencyGraph) -> bool:
             dependency = graph.dependency
+            # if already encountered we're done
             if dependency in children:
-                return False  # Nothing to do
+                return False
+            # if it's not a real file, then it's not a root
+            if not dependency.path.is_file():
```

Comment:
if `WorkspacePath` is used, then this will skip notebooks in the workspace, that are valid roots. can be do `not dependency.path.is_dir()` or something like it?

```
    def is_file(self) -> bool:
        """Return True if the path points to a file in Databricks Workspace."""
        try:
            return self._object_info.object_type == ObjectType.FILE
        except DatabricksError:
            return False
```

Reply:
which cyclic dependencies?

Reply:
see https://github.com/databrickslabs/ucx/pull/2236/files/0950ab1f809795fc7a086b1d4b04936ec83b9e26#r1697132861
https://github.com/databrickslabs/ucx/pull/2236/files#r1697135314
and
https://github.com/databrickslabs/ucx/pull/2236/files#r1697136495